src/fractgen.1
.\" Fractal Generator
.\" Copyright (C) 2003-2026 by Thomas Dreibholz
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact: thomas.dreibholz@gmail.com
.\"
.\" ###### Setup ############################################################
.Dd May 11, 2026
.Dt fractgen 1
.Os fractgen
.\" ###### Name #############################################################
.Sh NAME
.Nm fractgen
.Nd Fractal Generator
.\" ###### Synopsis #########################################################
.Sh SYNOPSIS
.Nm fractgen
.Op Ar file Ar ...
.Nm fractgen
.Op Fl h | Fl Fl help
.Nm fractgen
.Op Fl v | Fl Fl version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm fractgen
is a simple Qt\-based fractal generator program for Mandelbrot fractals. The
image size is only limited by virtual memory. It is possible to zoom into
images. Image parameters can be saved in and loaded from FractGen Settings
Files (FSF). These files are encoded in XML. While the standard
file extension is .fsf,
.Nm
detects the file type by reading its XML structure, meaning the extension is
not strictly required.
.br
The intention of this program is to generate graphics to be post\-processed
by other image tools, e.g. in order to generate nice screen backgrounds or
book covers.
.br
This is the Qt version of FractGen.
.Pp
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following argument may be provided:
.Bl -tag -width indent
.It Ar file Ar ...
Specifies input files to be loaded.
.It Fl h | Fl Fl help
Show arguments help and exit.
.It Fl v | Fl Fl version
Show version and exit.
.El
.\" ###### Examples #########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It fractgen \-h
.It fractgen \-v
.It fractgen fractgen.fsf
.It fractgen beta01.fsf beta02.fsf beta03.fsf
.It fractgen examples/alpha*.fsf
.El
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr clifractgen 1 ,
.Xr kfractgen 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of FractGen. The latest version of FractGen can be found on the FractGen Homepage at
.Lk https://www.nntb.no/\(tidreibh/fractalgenerator/ "FractGen Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit 0348d38f845a51182da822843d85563138446321
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 23:39:35 2026 +0200

    Packaging updates.

commit 3a6994945bb425807e2ba2d6143e983b55439697
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 23:14:27 2026 +0200

    Updated manpages.

commit d6508d530f3f8bb9a24705ecaa03b83d8243edfb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 23:14:14 2026 +0200

    Further parameter handling improvements.

commit 92e7a83945689d81953019ce32d67818b3a5443a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 23:02:49 2026 +0200

    Improved manpage.

commit 7f6c81df21c9d1bc79dab9a589bc7c66dcc1da88
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 22:43:26 2026 +0200

    Initialisation improvements, particularly now properly handling positional parameters.

commit 79733145888a7a1466cd10c447111ad441faabca
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 22:37:30 2026 +0200

    Minor documentation improvements.

commit 47668d078c804855d85306fb0a032fbf88c59b73
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 22:21:58 2026 +0200

    Improved documentation.

commit f778a95a9ceae8e0910d47cb8a68860e8bc10dc6
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 21:51:39 2026 +0200

    Improved manpages.

commit 4f0e83c5b3231f5fa373b455c86ec013c9769e90
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon May 11 21:48:45 2026 +0200

    Improved shell completions.

commit a6b6a32a776911a3a27d4cc4ac7a8a35672166cb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 27 11:53:13 2026 +0200

    Updated workflows.

commit d0a60170b1d01ff024526bd243adf87c945e4ba5
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 27 11:51:38 2026 +0200

    ShellCheck clean-up for shell completions.

commit 016a49ecf5cff25e29336df5c042fe06c20e8fc9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 27 11:49:34 2026 +0200

    Updated CI configuration

commit 8cdecef92fd164285f52dea4d71d2655ab89251a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 25 20:08:08 2026 +0200

    Updated build script.

commit 9694474e34f7ef8da0d0f6d30dddda683f4fd8e8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Sat Apr 25 20:07:34 2026 +0200

    Updated CI scripts

commit 28f3b94aa578d8c64771d3511769f501ea19c276
Merge: ea0c57a e7a9423
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 20 22:09:45 2026 +0200

    Merge branch 'master' of github.com:dreibh/fractgen

commit ea0c57acb953a64238f3af7435c9a65a85c92060
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 20 22:09:31 2026 +0200

    RPM spec fix.

commit e7a94231c7090df079011124ff97f24544584427
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 20 20:19:32 2026 +0200

    Updated FreeBSD port.

commit f616af1cc35e5df90d3ed3be774dc65f93f7de7e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 20 20:10:08 2026 +0200

    New release fractgen-3.0.8.

commit a0263910f5cd66c0093fee64d6c00901a92333b2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 20 09:47:22 2026 +0200

    Another dependency fix.

commit 813d92b7b140d516eea8364bde147d7aad45ffdd
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 20 09:38:24 2026 +0200

    Fixed typo.

commit b81dedd56fe71b4175fea4a8cce7f85069d86be3
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 20 09:28:20 2026 +0200

    Fixed Debian packaging conflict.

commit 757e335719d082abc18d37cdd92321c63bc8f62b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Apr 20 09:22:59 2026 +0200

    Updated CI scripts

commit fcbcc2cccb5a2e9bc9267adbbc5158b48b2c4774
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Mar 12 11:13:14 2026 +0100

    Updated URLs.

commit 2f6020570436313ae235d8ed8f6b69f1c0a3ce65
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Feb 27 09:27:56 2026 +0100

    Updated FreeBSD port.

commit 3b5313b096d3c10404763ab19fc6e30975f23b99
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Feb 27 08:56:10 2026 +0100

    New release fractgen-3.0.7.

commit 428e75acd0ea0122aac8e6d46aa5ae2daec2f47f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 26 13:37:37 2026 +0100

    Translation fix.

commit 51ef4fa7548410c1687e4b59a00a40ad59780e7b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 26 13:35:33 2026 +0100

    Improved error handling when saving a file.

commit 1b3dc06ccdc1acbce5aaf2f4e7b0af63e42caba0
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 26 12:13:08 2026 +0100

    Do not print misleading error if there are no translations for the current locale setting.

commit f74f1d9cced0e194bdd9f7c47296e4a1ed8687f3
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 26 12:01:37 2026 +0100

    Added AppStream metainfo.

commit 7e042730917c2b3930f4df35f5d748a90bf747b0
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Feb 26 11:32:07 2026 +0100

    Added metainfo files.

commit bdf3c5d86f3ffcd96fdabc72f5cd65e1bc4bfa9f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Dec 12 18:55:32 2025 +0100

    Updated documentation.

commit 6d20a496d1f3b5385a854a91a352fa7e77da4aa8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Dec 9 16:20:03 2025 +0100

    Updated CI scripts

commit 860def4cb99856f24bff6dc36bfe637f444dc1be
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Dec 9 16:19:40 2025 +0100

    CI workflow should be okay now.

commit a6076e7d03511af13b7508891d728ac6f56f8426
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Dec 9 15:30:13 2025 +0100

    One more try.

commit 280f2be21a8cb67fa1b3714e36b12ba81462a003
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Dec 9 15:10:50 2025 +0100

    Another try.

commit 563b9934db239c0b722d2d6d4fb1900345093f52
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Dec 9 14:55:47 2025 +0100

    Yet another try.

commit 3e94cbaefb59012b056b0a3f562f12e1960eee9a

src/fractgen.cc
/* ==========================================================================
 * ====                   FRACTAL GRAPHICS GENERATOR                     ====
 * ==========================================================================
 *
 * Copyright (C) 2003-2026 by Thomas Dreibholz
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Contact: thomas.dreibholz@gmail.com
 */

#include "fractalgenerator.h"
#include "package-version.h"

#include <QCommandLineParser>
#include <QFile>
#include <QtGlobal>
#include <QTranslator>
#include <QtWidgets/QApplication>


// ###### Main program ######################################################
int main(int argc, char *argv[])
{
   // ====== Initialise =====================================================
   QApplication application(argc, argv);
#if QT_VERSION < 0x060000
   application.setAttribute(Qt::AA_UseHighDpiPixmaps, true);
#endif
   QTranslator applicationTranslator;
   if(!applicationTranslator.load(QLatin1String("fractgen_") + QLocale::system().name())) {
      if(!applicationTranslator.load(QLatin1String("fractgen_") + QLocale::system().name(),
                                     QStringLiteral(FRACTGEN_INSTALL_PREFIX) + QStringLiteral("/share/fractgen"))) {
         // No translations for this locale -> using built-in strings.
      }
   }
   application.installTranslator(&applicationTranslator);
#ifdef WITH_KDE
   QCoreApplication::setApplicationName(QStringLiteral("kfractgen"));
#else
   QCoreApplication::setApplicationName(QStringLiteral("fractgen"));
#endif
   QCoreApplication::setApplicationVersion(QStringLiteral(FRACTGEN_VERSION));

   // ====== Parse command-line arguments ===================================
   QCommandLineParser parser;
   parser.setApplicationDescription(QStringLiteral("Fractal Generator"));
   parser.addHelpOption();
   parser.addVersionOption();
   parser.addPositionalArgument(QStringLiteral("input_file"),
                                QStringLiteral("Input file."),
                                QStringLiteral("[input_file ...]"));
   parser.process(application);

   // ====== Open files given as arguments ==================================
   bool  openedAnyFile    = false;
   const QStringList args = parser.positionalArguments();
   for(const QString& fileName : args) {
      if(QFile::exists(fileName)) {
         // Open file provided by argument ...
         FractalGeneratorApp* fractalGeneratorApp =
            new FractalGeneratorApp(nullptr, fileName);
         Q_CHECK_PTR(fractalGeneratorApp);
         fractalGeneratorApp->show();
         openedAnyFile = true;
      }
      else {
         qWarning("WARNING: File %s does not exist!", qPrintable(fileName));
      }
   }

   // ====== Start the application ==========================================
   if(!openedAnyFile) {
      // Start new image, if no file has been opened.
      FractalGeneratorApp* fractalGeneratorApp =
         new FractalGeneratorApp(nullptr);
      Q_CHECK_PTR(fractalGeneratorApp);
      fractalGeneratorApp->show();
   }

   return application.exec();
}
